tests/xml/common/src/main/java/sparklr/common/AbstractRefreshTokenSupportTests.java

```diff
@@ -32,13 +32,36 @@ public void testHappyDay() throws Exception {
 
 		// now use the refresh token to get a new access token.
 		assertNotNull(accessToken.getRefreshToken());
-		OAuth2AccessToken newAccessToken = refreshAccessToken(accessToken.getRefreshToken().getValue());
+		OAuth2AccessToken newAccessToken = refreshAccessToken(accessToken.getRefreshToken().getValue(),
+			"my-trusted-client", "read");
 		assertFalse(newAccessToken.getValue().equals(accessToken.getValue()));
 
 		verifyAccessTokens(accessToken, newAccessToken);
 
 	}
 
+	/**
+	 * Tests that refreshing an access token specifying no scopes for a client
+	 * with unlimited scopes returns an access token with the originally
+	 * granted scopes.
+	 */
+	@Test
+	public void testGrantedScope() throws Exception {
+
+		OAuth2AccessToken accessToken = getAccessToken("read write", "my-client-with-empty-scopes");
```

Comment:
This call should fail given that `my-client-with-empty-scopes` is configured with `null` scopes and the request contains `requested` scopes. The Authorization Server should only grant scopes if the `requested` scopes **are a subset** of the `registered` scopes.

Reply:
I'm not sure this is right -- from the [docs](https://projects.spring.io/spring-security-oauth/docs/oauth2.html#configuring-client-details):

> `scope`: The scope to which the client is limited. If scope is undefined or empty (the default) the client is not limited by scope.

Thus a client with empty scopes accepts **any** requested scopes.

Reply:
Please refer to the [spec](https://tools.ietf.org/html/rfc6749#section-6)

>  scope
         OPTIONAL.  The scope of the access request as described by
         Section 3.3.  **The requested scope MUST NOT include any scope
         not originally granted by the resource owner**, and if omitted is
         treated as equal to the scope originally granted by the
         resource owner.

Take note of the bold text. Given this, the test needs to be changed. Makes sense?

Reply:
I think we've crossed wires here. The line that this thread is against obtains the initial access token, not the subsequent refresh that you linked to in the spec. That occurs on [line 55](https://github.com/spring-projects/spring-security-oauth/pull/1697/files#diff-9f608bad59b62b21a9b3b2f0e2bb5772R55) where the scopes are empty.

Reply:
My apologies. I crossed the wires. However, this test still does not make sense to me. The client is requesting `read write` scopes and is granted these scopes even though it doesn't have these scopes configured in it's registration. So basically, the client can request any scope, including a higher-privilege scope, e.g. `admin`.

If the client doesn't request any scopes, than it's up to the provider to return a default set of scopes OR empty scope.

If the client requests one or more scopes, than the provider should validate each requested scope to ensure it's also configured (or allowed to be requested) with the client.

Makes sense?

Reply:
No probs, glad to be on the same page. Regarding a client registered with no scopes, the spec doesn't cover registration but the [Spring docs](https://projects.spring.io/spring-security-oauth/docs/oauth2.html#configuring-client-details) state:

> `scope`: The scope to which the client is limited. If scope is undefined or empty (the default) the client is not limited by scope.

I read this as: *if the client's registered scopes are empty then it will support arbitrary scopes*.

[`DefaultOAuth2RequestValidator`](https://github.com/spring-projects/spring-security-oauth/blob/8eedb354ae0015e0f33f0c3d958d9ac84d010721/spring-security-oauth2/src/main/java/org/springframework/security/oauth2/provider/request/DefaultOAuth2RequestValidator.java#L29) embodies this by not validating scopes when the client hasn't been registered with any:

```java
if (clientScopes != null && !clientScopes.isEmpty()) {
    for (String scope : requestScopes) {
        if (!clientScopes.contains(scope)) {
            throw new InvalidScopeException("Invalid scope: " + scope, clientScopes);
        }
    }
}
```

If this is deemed to be incorrect behaviour then that would be a separate issue which would negate the need for this PR. What do you think?

Reply:
> if the client's registered scopes are empty then it will support arbitrary scopes.

No, this is not correct. A client cannot request any arbitrary scope and automatically be granted it. A client can only request a scope(s) that it has been allowed to request based on it's client configuration/registration with the provider.

If the client has empty scopes configured than it is not limited by scope and instead limited by some other authority information associated in the token. However, this also means that the client cannot request any scopes, given it's not configured to allow to request any scopes.

I want to point out that the validation of scopes is different during a `refresh_token` grant (**re-authorization**) compared to the initial **authorization** request initiated via the `authorization_code` or `client_credentials` grant.

When authorization is initially requested, using `authorization_code` or `client_credentials` grant, the requested scopes must be validated against the configured/registered scopes of the client. The requested scopes must be containing within the configured/registered scopes of the client. If the client is not configured with any scopes than the client cannot request any scopes either.

When re-authorization is requested, using `refresh_token` grant, the requested scopes must be validated against the previously granted scopes from the authorization request. This is the main difference in validation compared to the initial authorization where the configured/registered scopes are used instead for validating against.

Reply:
@markhobson I haven't heard back from you in a while. Have you resolved the issue you were having in a different way? Or are you still planning on addressing it here? Please let me know so we can address your issue or close this PR.

Reply:
Hi @jgrandja, thanks for the reminder. I don't believe this is causing me any issues now that I've registered my client with scopes. My conclusion was:

> I believe this whole issue can be worked around if the client is registered with scopes. In my case it wasn't, so coupled with the empty scope refresh request it highlighted this edge case.

I'm happy to close this issue, although we may wish to clarify the [Spring docs](https://projects.spring.io/spring-security-oauth/docs/oauth2.html#configuring-client-details) regarding registering a client with empty scopes to something like you wrote above, e.g.:

> If scope is undefined or empty (the default) the client is not limited by scope and instead limited by some other authority information associated in the token.

Also, the exception message I initially received when refreshing a token without scopes, for a client registered with empty scopes, could perhaps be improved, i.e.:

> InvalidScopeException: Empty scope (either the client or the user is not allowed the requested scopes)

Thanks for your time working through this.